tools/state-parts-dump/src/cli.rs

```diff
@@ -0,0 +1,167 @@
+        let part_file_names = external.list_state_parts(self.shard_id, &directory_path).await?;
+        // TODO: return error instead of assert!()
+        assert!(!part_file_names.is_empty(), "Number of state parts is 0");
+        let num_parts = part_file_names.len() as u64;
+        let total_required_parts = get_num_parts_from_filename(&part_file_names[0]).ok_or(anyhow!("get_num_parts_from_filename returned None"))?;
+        
+        crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_PARTS_TOTAL.with_label_values(&[&self.shard_id.to_string()]).set(total_required_parts as i64);
+        crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_PARTS_DUMPED.with_label_values(&[&self.shard_id.to_string()]).set(num_parts as i64);
+
+        if num_parts < total_required_parts {
+            println!("total state parts required: {} < number of parts already dumped: {}, waiting for all parts to be dumped", total_required_parts, num_parts);
+            return Ok(StatePartsDumpCheckStatus::WaitingForParts { epoch_height: self.epoch_height })
+        } else if num_parts > total_required_parts{
+            println!("total state parts required: {} > number of parts already dumped: {}, there are more dumped parts than total required, something is seriously wrong", total_required_parts, num_parts);
+            return Ok(StatePartsDumpCheckStatus::Done { epoch_height: self.epoch_height })
+        }
+
+        // TODO: run in parallel with rayon
+        for part_id in 0..total_required_parts {
+            if part_id == 0 {
+                crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_PARTS_VALID.with_label_values(&[&self.shard_id.to_string()]).set(0);
+                crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_PARTS_INVALID.with_label_values(&[&self.shard_id.to_string()]).set(0);
+            }
+            let location = external_storage_location(
+                &self.chain_id,
+                &self.epoch_id,
+                self.epoch_height,
+                self.shard_id,
+                part_id,
+                num_parts,
+            );
+            let part = external.get_part(self.shard_id, &location).await?;
+            let is_part_valid = validate_state_part(
+                &self.state_root,
+                PartId::new(part_id, num_parts),
+                &part
+            );
+            if is_part_valid {
+                crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_PARTS_VALID.with_label_values(&[&self.shard_id.to_string()]).inc();
+                println!("part {} is valid", part_id);
+            } else {
+                crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_PARTS_INVALID.with_label_values(&[&self.shard_id.to_string()]).inc();
+                println!("part {} is invalid", part_id);
+            }
+        };
+        Ok(StatePartsDumpCheckStatus::Done { epoch_height: self.epoch_height })
+    }
+}
+
+
+pub fn create_external_connection(
```

Comment:
I've seen this function before somewhere :eyes:

Reply:
This was in state sync code but I removed some unnecessary arguments thus a new one here